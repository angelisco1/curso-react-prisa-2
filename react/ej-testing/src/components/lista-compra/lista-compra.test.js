import React from 'react';
import { shallow, mount } from 'enzyme';
import ListaCompra from './lista-compra';
import Producto from './producto';

describe('ListaCompra component', () => {
  let wrapper;
  const productos = [
    { id: 0, nombre: 'One Plus 5T', checked: true },
    { id: 1, nombre: 'Funda One Plus', checked: false },
    { id: 2, nombre: 'SD 64GB', checked: false }
  ];
  beforeEach(() => {
    wrapper = shallow(<ListaCompra logueado={ true } productos={ productos } />)
  });
  
  it('should not render Producto components and a Login button if not logueado', () => {
    wrapper.setState({ productos: [], input: '' })
    wrapper.setProps({ logueado: false })
    expect(wrapper.find(Producto)).toHaveLength(0);
    expect(wrapper.find('button').text()).toBe('Login');
  });
    
  it('should render three Producto components and a Logout button if logueado', () => {
    expect(wrapper.find(Producto)).toHaveLength(3);
    expect(wrapper.find('button').text()).toBe('Logout');
  });
  
  it('should render a message if there is no products and logueado', () => {
    wrapper.setState({ productos: [] });
    expect(wrapper.find(Producto)).toHaveLength(0);
    expect(wrapper.find('p').text()).toBe('Lista vacia');
  });
  
  it('should render the product name if logueado (using mount instead of shallow)', () => {
    wrapper = mount(<ListaCompra logueado={true} productos={[{ id: 1, nombre: 'One Plus 5T' }]} />);
    const name = wrapper.find('span').first().text();
    expect(name).toEqual('One Plus 5T');
  });

  it('should match its empty snapshot', () => {
    wrapper = shallow(<ListaCompra logueado={ false } />);
    expect(wrapper).toMatchSnapshot();
  });
  
  it('should match its snapshot with products and logueado', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  it('should match its snapshot with empty products and logueado', () => {
    wrapper.setState({
      productos: []
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('should check a product', () => {    
    wrapper = mount(<ListaCompra logueado={true} productos={productos} />)
    // Aqui el primer producto est치 checked
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Producto).first().props().checked).toBe(true);
    // Pulsamos en el primer producto para cambiar el valor de su checked
    const product = wrapper.find('div.producto').first();
    product.simulate('click');
    // Aqui el primer producto no est치 checked
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Producto).first().props().checked).toBe(false);
  });

  it('should add a new product to the list', () => {    
    wrapper = mount(<ListaCompra logueado={true} productos={productos} />)
    // Aqui el primer producto est치 checked
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Producto)).toHaveLength(3);
    // Pulsamos en el primer producto para cambiar el valor de su checked
    const form = wrapper.find('form');
    const input = wrapper.find('input');
    input.simulate('change', { target: { value: 'un producto nuevo' } });
    form.simulate('submit');
    // Aqui se ha a침adido el producto nuevo
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Producto)).toHaveLength(4);
  });
});