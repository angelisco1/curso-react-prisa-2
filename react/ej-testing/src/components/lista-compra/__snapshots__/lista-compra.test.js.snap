// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListaCompra component should add a new product to the list 1`] = `
<ListaCompra
  logueado={true}
  productos={
    Array [
      Object {
        "checked": false,
        "id": 0,
        "nombre": "One Plus 5T",
      },
      Object {
        "checked": false,
        "id": 1,
        "nombre": "Funda One Plus",
      },
      Object {
        "checked": false,
        "id": 2,
        "nombre": "SD 64GB",
      },
    ]
  }
>
  <div>
    <h1>
      Lista compra
    </h1>
    <Producto
      checked={false}
      key="0"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <span>
          One Plus 5T
        </span>
      </div>
    </Producto>
    <Producto
      checked={false}
      key="1"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <span>
          Funda One Plus
        </span>
      </div>
    </Producto>
    <Producto
      checked={false}
      key="2"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <span>
          SD 64GB
        </span>
      </div>
    </Producto>
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </form>
    <button
      type="button"
    >
      Logout
    </button>
  </div>
</ListaCompra>
`;

exports[`ListaCompra component should add a new product to the list 2`] = `
<ListaCompra
  logueado={true}
  productos={
    Array [
      Object {
        "checked": false,
        "id": 0,
        "nombre": "One Plus 5T",
      },
      Object {
        "checked": false,
        "id": 1,
        "nombre": "Funda One Plus",
      },
      Object {
        "checked": false,
        "id": 2,
        "nombre": "SD 64GB",
      },
    ]
  }
>
  <div>
    <h1>
      Lista compra
    </h1>
    <Producto
      checked={false}
      key="0"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <span>
          One Plus 5T
        </span>
      </div>
    </Producto>
    <Producto
      checked={false}
      key="1"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <span>
          Funda One Plus
        </span>
      </div>
    </Producto>
    <Producto
      checked={false}
      key="2"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <span>
          SD 64GB
        </span>
      </div>
    </Producto>
    <Producto
      checked={false}
      key="3"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <span>
          un producto nuevo
        </span>
      </div>
    </Producto>
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </form>
    <button
      type="button"
    >
      Logout
    </button>
  </div>
</ListaCompra>
`;

exports[`ListaCompra component should check a product 1`] = `
<ListaCompra
  logueado={true}
  productos={
    Array [
      Object {
        "checked": true,
        "id": 0,
        "nombre": "One Plus 5T",
      },
      Object {
        "checked": false,
        "id": 1,
        "nombre": "Funda One Plus",
      },
      Object {
        "checked": false,
        "id": 2,
        "nombre": "SD 64GB",
      },
    ]
  }
>
  <div>
    <h1>
      Lista compra
    </h1>
    <Producto
      checked={true}
      key="0"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "line-through",
          }
        }
      >
        <span>
          One Plus 5T
        </span>
      </div>
    </Producto>
    <Producto
      checked={false}
      key="1"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <span>
          Funda One Plus
        </span>
      </div>
    </Producto>
    <Producto
      checked={false}
      key="2"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <span>
          SD 64GB
        </span>
      </div>
    </Producto>
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </form>
    <button
      type="button"
    >
      Logout
    </button>
  </div>
</ListaCompra>
`;

exports[`ListaCompra component should check a product 2`] = `
<ListaCompra
  logueado={true}
  productos={
    Array [
      Object {
        "checked": false,
        "id": 0,
        "nombre": "One Plus 5T",
      },
      Object {
        "checked": false,
        "id": 1,
        "nombre": "Funda One Plus",
      },
      Object {
        "checked": false,
        "id": 2,
        "nombre": "SD 64GB",
      },
    ]
  }
>
  <div>
    <h1>
      Lista compra
    </h1>
    <Producto
      checked={false}
      key="0"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <span>
          One Plus 5T
        </span>
      </div>
    </Producto>
    <Producto
      checked={false}
      key="1"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <span>
          Funda One Plus
        </span>
      </div>
    </Producto>
    <Producto
      checked={false}
      key="2"
      onHandleClick={[Function]}
    >
      <div
        className="producto"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <span>
          SD 64GB
        </span>
      </div>
    </Producto>
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </form>
    <button
      type="button"
    >
      Logout
    </button>
  </div>
</ListaCompra>
`;

exports[`ListaCompra component should match its empty snapshot 1`] = `
<div>
  <h1>
    No estas logueado
  </h1>
  <button
    type="button"
  >
    Login
  </button>
</div>
`;

exports[`ListaCompra component should match its snapshot with empty products and logueado 1`] = `
<div>
  <h1>
    Lista compra
  </h1>
  <p>
    Lista vacia
  </p>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </form>
  <button
    type="button"
  >
    Logout
  </button>
</div>
`;

exports[`ListaCompra component should match its snapshot with products and logueado 1`] = `
<div>
  <h1>
    Lista compra
  </h1>
  <Producto
    checked={true}
    key="0"
    onHandleClick={[Function]}
  >
    One Plus 5T
  </Producto>
  <Producto
    checked={false}
    key="1"
    onHandleClick={[Function]}
  >
    Funda One Plus
  </Producto>
  <Producto
    checked={false}
    key="2"
    onHandleClick={[Function]}
  >
    SD 64GB
  </Producto>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </form>
  <button
    type="button"
  >
    Logout
  </button>
</div>
`;
